two's-complement logic operations on BigInt

tspiteri opened the pull request:
This commit adds the following impls, giving results equivalent to what would be expected with two's complement.

* `impl Not for BigInt`
* `impl BitAnd<&BigInt> for BigInt`, `impl BitAndAssign<&BigInt> for BigInt`
* `impl BitOr<&BigInt> for BigInt`, `impl BitOrAssign<&BigInt> for BigInt`
* `impl BitXor<&BigInt> for BigInt`, `impl BitXorAssign<&BigInt> for BigInt`

I also added some tests. There is most probably some room for optimization, but the bitwise operations all operate in place using a single pass.

I did not add other implementations such as `impl BitAnd<BigInt> for BigInt`, those can be done later.

Edit: Fixes #13

cuviper commented on src/bigint.rs:
```diff
@@ -183,6 +185,390 @@ impl fmt::UpperHex for BigInt {
+
+// -1 ^ +ff = ...ffff ^ ...00ff = ...ff00 = +100
+// -ff ^ +1 = ...ff01 ^ ...0001 = ...ff00 = +100
+// answer is pos, has length of longest with a possible carry
```
answer is *neg*, but I think you got that right in practice.

tspiteri commented:
I addressed your points in the first commit.

In the second commit I added tests on `i64` values within ±3 of interesting values (`i64::MIN`, `-u32::MAX`, etc.). All logic operations are tested on all combinations with both `i64` and `BigInt`, and the answers compared. That found a couple of bugs I fixed.

tspiteri commented:
I've added a third commit which makes the inline docs more helpful. I also made the use of the `debug_assertions` more consistent. I think this makes the code easier to understand.

tspiteri commented on src/bigint.rs:
```diff
@@ -179,6 +180,454 @@ impl fmt::UpperHex for BigInt {
+    fn not(mut self) -> BigInt {
+        match self.sign {
+            Sign::NoSign | Sign::Plus => {
+                self.data += 1 as BigDigit;
```
You guessed right :)

tspiteri commented:
Commit 1 is now the implementation and commit 2 is the tests.

Commit 3 introduces a crate-visible `IntDigits` trait so that macros that used to use `data.len()` or `data.capacity()` can work for both `BigUint` and `BigInt`, and I also used it for `BigInt` to access `BigUint::normalize`. I made `<BigUint as IntDigits>::normalize` simply call the inherent method `BigUint::normalize`, maybe the implementation could move into the trait method and the inherent method can simply be removed?

Commit 4 implements ref/val forwarding for the bitwise ops of `BigInt` (depends on commit 3).

tspiteri commented on src/biguint.rs:
```diff
@@ -1686,6 +1686,24 @@ pub fn trailing_zeros(u: &BigUint) -> Option<usize> {
         .map(|(i, digit)| i * big_digit::BITS + digit.trailing_zeros() as usize)
 }
 
+// needed since we cannot set visibility of `BigUint::data` to `pub(crate)`
```
I'm not sure I understand this point. I have done something in the new commit 3 (crate-visible IntDigits) that maybe addresses this, but I'm not sure if that's what you had in mind. Anyway, I think further changes here can be independent of this PR.

cuviper replied:
I meant that if we moved the struct definition to the crate root, then everything could just access the `data` field directly.  The `digits_mut` function especially was defeating any pretense of privacy within the crate anyway.

But now with your trait, I see how this is helpful to present a common interface for the macros, so this solution seems fine.

cuviper commented:
I took the small liberty of adding reference `Not`, but otherwise this quite comprehensive.  Thanks a lot!

bors r+
